lib/SILGen/SILGenBuiltin.cpp

```diff
@@ -1042,28 +1042,60 @@ static ManagedValue emitBuiltinTypeTrait(SILGenFunction &SGF,
 }
 
 // SWIFT_ENABLE_TENSORFLOW
-static ManagedValue emitBuiltinAutoDiffGetJVP(SILGenFunction &SGF,
-                                              SILLocation loc,
-                                              SubstitutionMap substitutions,
-                                              Expr *argument,
-                                              SGFContext C) {
-  auto argVal = SGF.emitRValue(argument);
-  auto jvp = SGF.getBuilder().createAutoDiffFunctionExtract(
-      loc, AutoDiffFunctionExtractee::JVP, /*differentiationOrder*/ 1,
-      std::move(argVal).forwardAsSingleValue(SGF, loc));
-  return ManagedValue::forUnmanaged(jvp);
+static ManagedValue emitBuiltinAutoDiffApplyAssociatedFunction(
```

Comment:
Ah, I think I understand part of what is making this difficult. It is unfortunate that builtins emit `ManagedValue`s rather than `RValue`s. `RValue`s are exploded tuples that can express things like `($*U, $Pullback)` where one element is an address and the other isn't. `ManagedValue`s contain a single value so the best we can do if the first element of the tuple is address-only is to make a `ManagedValue` of type `$*(U, Pullback)`.

This causes lots of inefficiency:
* `SGF.getBufferForExprResult` allocates space on the stack for the tuple `$*(U, Pullback)` because it can't use the indirect result argument `$*U` that often exits. Then the first element of that tuple later gets copied to the indirect result argument.
* The pullback must be stored into the `$*(U, Pullback)` and then immediately loaded by its user.

Anyways, it's probably best to keep it like this for now for expediency. I wonder how hard it would be to make builtins that emit `RValue`s though...

Reply:
Yeah, it's complicated and I need to understand SILGen more. After running through optimizations, the second element of the buffer does become dead code.

```swift
bb0(%0 : $*U, %1 : $@autodiff @noescape @callee_guaranteed (@in_guaranteed T) -> @out U, %2 : $*T):
  debug_value %1 : $@autodiff @noescape @callee_guaranteed (@in_guaranteed T) -> @out U, let, name "f", argno 1 // id: %3
  debug_value_addr %2 : $*T, let, name "x", argno 2 // id: %4
  %5 = autodiff_function_extract [jvp] [order 1] %1 : $@autodiff @noescape @callee_guaranteed (@in_guaranteed T) -> @out U // user: %8
  %6 = alloc_stack $(U, @callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector) // users: %12, %9, %7
  %7 = tuple_element_addr %6 : $*(U, @callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector), 0 // users: %11, %8
  %8 = apply %5(%7, %2) : $@noescape @callee_guaranteed (@in_guaranteed T) -> (@out U, @owned @callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector) // users: %13, %10
  %9 = tuple_element_addr %6 : $*(U, @callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector), 1 // user: %10
  store %8 to %9 : $*@callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector // id: %10
  copy_addr [take] %7 to [initialization] %0 : $*U // id: %11
  dealloc_stack %6 : $*(U, @callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector) // id: %12
  return %8 : $@callee_guaranteed (@in_guaranteed T) -> @out U.TangentVector // id: %13
} // end sil function '$s13core_builtins8evaldiffyq__13TangentVectorQy_xctq_xXE_xts14DifferentiableRzACQzRszsAER_r0_lF'
```